[MRG] Add quantile regression (#9978)
scikit-learn/scikit-learn: doc/modules/linear_model.rst

squares implementation with weights given to each sample on the basis of how much the residual is
greater than a certain threshold.

.. _quantile_regression:

Quantile Regression
===================

Quantile regression estimates median or other quantiles of :math:`y` conditional on :math:`X`,

>     while OLS estimates conditional mean.

Reviewer: Please remove spaces in the beginning of the line, see rendered docs at https://20523-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/linear_model.html#quantile-regression

.. math::
    \underset{w}{min\,} { \frac{1}{n_{samples}} L_q (y - X w) + \alpha \rho ||w||_1 + \alpha(1-\rho) ||w||_2 ^ 2}

where

.. math::
    \L_q(t) =
    \begin{cases}
        q t, & t > 0, \\
        0,    & t = 0, \\
        (1-q) t, & t < 0
    \end{cases}

and :math:`q \in (0, 1)` is the quantile to be estimated.

Quantile regression may be useful if one is interested in predicting an interval
instead of point prediction. Sometimes prediction interval is calculated based on
assumption that prediction error is distributed normally with zero mean and constant variance.
Quantile regression provides sensible prediction intervals even for errors with non-constant
(but predictable) variance or non-normal distribution.

.. figure:: /auto_examples/linear_model/images/sphx_glr_plot_quantile_regression_001.png
   :target: ../auto_examples/linear_model/plot_quantile_regression.html
   :align: center
   :scale: 50%

Another possible advantage of quantile regression over OLS is its robustness
to outliers, because it is only sign of an error that influences estimated
coefficients, not its absolute value.

Quantile loss function can be used with models other than linear.

> :class:`GradientBoostingRegressor` also has an option to predict conditional quantiles.

Reviewer: Please provide more details, about how this could be achieved (the exact parameters are described in the parent issue I think).

like :class:`ElasticNet`. It solves

.. math::
    \underset{w}{min\,} { \frac{1}{n_{samples}} L_q (y - X w) + \alpha \rho ||w||_1 + \alpha(1-\rho) ||w||_2 ^ 2}

where

.. math::
    \L_q(t) =
    \begin{cases}
        q t, & t > 0, \\
        0,    & t = 0, \\
        (1-q) t, & t < 0
    \end{cases}

and :math:`q \in (0, 1)` is the quantile to be estimated.

Quantile regression may be useful if one is interested in predicting an interval
instead of point prediction. Sometimes prediction interval is calculated based on
assumption that prediction error is distributed normally with zero mean and constant variance.
Quantile regression provides sensible prediction intervals even for errors with non-constant
(but predictable) variance or non-normal distribution.

.. figure:: /auto_examples/linear_model/images/sphx_glr_plot_quantile_regression_001.png
   :target: ../auto_examples/linear_model/plot_quantile_regression.html
   :align: center
   :scale: 50%

Another possible advantage of quantile regression over OLS is its robustness
to outliers, because it is only sign of an error that influences estimated

> coefficients, not its absolute value.

Reviewer 2: shouldn't this be "it is the value [...] rather than the squared error" ? "only the sign of an error" would correspond to L0 loss

Author: The model parameters are updated by gradient descent, and the gradient of L1 loss is the sign of the error, while the gradient of L2 loss is twice the error.

The explanation in term of gradients seems more intuitive to me, because it explicitly shows that with L1 loss it does not matter how large the error of a particular data point is - the resulting parameter estimates will be the same, as long as the sign of the error does not change.

squares implementation with weights given to each sample on the basis of how much the residual is
greater than a certain threshold.

.. _quantile_regression:

Quantile Regression
===================

Quantile regression estimates the median or other quantiles of :math:`y`
conditional on :math:`X`, while OLS estimates the conditional mean.

As a linear model, the :class:`QuantileRegressor` gives linear predictions :math:`\hat{y}(w, x) = Xw` for the specified ``quantile`` :math:`q \in (0, 1)`. The weights or coefficients :math:`w` are then found by the following minimization problem:

.. math::

>     \min_{w} {\frac{1}{n_{\text{samples}}} PB_q(y - X w) + \alpha ||w||_1}

Reviewer 3: ```suggestion
    \min_{w} {\frac{1}{n_{\text{samples}}} \sum_i PB_q(y_i - X_i w) + \alpha ||w||_1}
```

This consists of the pinball loss (also known as linear loss),
see also :class:`~sklearn.metrics.mean_pinball_loss`,

.. math::

>     \PB_q(t) = q \max(t, 0) + (1 - q) \max(-t, 0) =

Reviewer 3: ```suggestion
    PB_q(t) = q \max(t, 0) + (1 - q) \max(-t, 0) =
```

    \begin{cases}
        q t, & t > 0, \\
        0,    & t = 0, \\
        (1-q) t, & t < 0
    \end{cases}

and the L1 penalty, similar to the :class:`Lasso`.

As the pinball loss is only linear in the residuals, quantile regression

> is much morerobust to outliers than estimation of the mean by the squared error.

Reviewer: ```suggestion
is much robust to outliers than estimation of the mean by the squared error.
```

squares implementation with weights given to each sample on the basis of how much the residual is
greater than a certain threshold.

.. _quantile_regression:

Quantile Regression
===================

Quantile regression estimates median or other quantiles of :math:`y`
conditional on :math:`X`, while OLS estimates conditional mean.

> The :class:`QuantileRegressor` applies linear loss to all samples.
> Somewhat related, the :class:`HuberRegressor` applies
> linear penalty to a small fraction of outliers and quadratic loss to the rest
> of observations. :class:`QuantileRegressor` also supports L1
> regularization, like :class:`Lasso`. It solves
> 
> .. math::
>     \underset{w}{min\,} { \frac{1}{n_{samples}} L_q (y - X w) + \alpha ||w||_1}
> 
> where
> 
> .. math::
>     \L_q(t) =
>     \begin{cases}
>         q t, & t > 0, \\
>         0,    & t = 0, \\
>         (1-q) t, & t < 0
>     \end{cases}
> 
> and :math:`q \in (0, 1)` is the quantile to be estimated.

Reviewer 3: ```suggestion
As a linear model, the :class:`QuantileRegressor` gives linear predictions :math:`\hat{y}(w, x) = Xw` for the specified ``quantile`` :math:`q \in (0, 1)`. The weights or coefficients :math:`w` are then found by the following minimization problem:

.. math::
    \min_{w} {\frac{1}{n_{\text{samples}}} PB_q(y - X w) + \alpha ||w||_1}

This consists of the pinball loss (also known as linear loss),
see also :class:`~sklearn.metrics.mean_pinball_loss`,

.. math::
    \PB_q(t) = q \max(t, 0) + (1 - q) \max(-t, 0) =
    \begin{cases}
        q t, & t > 0, \\
        0,    & t = 0, \\
        (1-q) t, & t < 0
    \end{cases}

and the L1 penalty, similar to the :class:`Lasso`.

As the pinball loss is only linear in the residuals, quantile regression
is much morerobust to outliers than estimation of the mean by the squared error.
Somewhat in between is the :class:`HuberRegressor`.
```

Quantile regression may be useful if one is interested in predicting an interval

> instead of point prediction. Sometimes prediction interval is calculated based

Reviewer 3: ```suggestion
instead of point prediction. Sometimes, prediction intervals are calculated based
```

on the assumption that prediction error is distributed normally with zero mean
and constant variance. Quantile regression provides sensible prediction
intervals even for errors with non-constant (but predictable) variance or
non-normal distribution.

.. figure:: /auto_examples/linear_model/images/sphx_glr_plot_quantile_regression_001.png
   :target: ../auto_examples/linear_model/plot_quantile_regression.html
   :align: center
   :scale: 50%

> Another advantage of quantile regression over OLS is its robustness
> to outliers, because it is only the sign of an error that influences estimated
> coefficients, not its absolute value.
> 

Reviewer 3: ```suggestion
```

> Quantile loss function can be used with models other than linear. For example,

Reviewer 3: ```suggestion
Based on minimizing the pinball loss, conditional quantiles can also be estimated by models other than linear models. For example,
```
